Add VDAF mapping for aggregation flow (#223)
ietf-wg-ppm/ppm-specification: draft-gpew-priv-ppm.md

| unrecognizedTask        | An endpoint received a message with an unknown task ID. |
| outdatedConfig          | The message was generated using an outdated configuration. |
| batchInvalid            | A collect or aggregate-share request was made with invalid batch parameters. |

> | invalidReportShare      | The report share was invalid and could not be processed. |

Reviewer: This error type doesn't seem to be used anywhere.

Author: Leftover 👍

The leader and helpers then run the aggregate initialization flow to accomplish two tasks:

1. Recover and determine which input report shares are invalid.
1. For each valid report share, initialize the VDAF preparation process.

An invalid report share is marked with one of the following errors:

~~~
enum {
  batch-collected(0),
  report-replayed(1),
  report-dropped(2),
  hpke-unknown-config-id(3),
  hpke-decrypt-error(4),
  vdaf-prep-error(5),
} ReportShareError;
~~~

#### Leader Initialization

The leader begins the aggregate initialization phase with the set of candidate report
shares as follows:

1. Decrypt the input share for each report share as described in {{input-share-decryption}}.
1. Check that the resulting input share is valid as described in {{input-share-batch-validation}}.
1. Initialize VDAF preparation as described in {{input-share-prep}}.

If any step yields an invalid report share, the leader removes the report share from
the set of candidate reports. Once the leader has initialized this state for all valid
candidate report shares, it then creates an AggregateInitReq message for each helper to

> initialize the preparation of this candidate set. This message is structured as follows:

Reviewer 2: Does this have to be done serially? Could the leader instead send all of the shares and then discover later that one is bogus? Surely yes, because the proof could fail.

Author: Yep, it can. I added a note to that effect.

~~~
struct {
  Nonce nonce;
  Extension extensions<0..2^16-1>;
  HpkeCiphertext encrypted_input_share;
} ReportShare;

struct {
  TaskID task_id;
  opaque agg_param<0..2^16-1>;
  opaque helper_state<0..2^16>;
  ReportShare report_shares<1..2^16-1>;
} AggregateInitReq;
~~~

The `nonce` and `extensions` fields of each ReportShare match that in the Report
uploaded by the client. The `encrypted_input_share` field is the `HpkeCiphertext`
whose index in `Report.encrypted_input_shares` is equal to the index of the aggregator

> in the task's `aggregator_endpoints`. The `agg_param` field is an opaque, VDAF-specific

Reviewer 2: ...to which the share is being sent...

aggregation parameter. The `helper_state` parameter contains the helper's state.
This is an optional parameter of an aggregate request that the helper can use to carry
state across requests and across aggregate flows.

Let `[aggregator]` denote the helper's API endpoint. The leader sends a POST
request to `[aggregator]/aggregate` with its AggregateInitReq message as
the payload. The media type is "message/ppm-aggregate-init-req".

#### Helper Initialization

Each helper begins their portion of the aggregate initialization phase with the set
of candidate report shares obtained in an `AggregateInitReq` message from the leader.
It attempts to recover and validate the corresponding input shares similar to the leader,
and eventually returns a response to the leader carrying a VDAF-specific message for each

> report share.

Reviewer 2: I wonder if it would be better to factor out the common behaviors for all aggregators including the leader.

Author: I think this done right now. The commonality is these steps, each of which have their own sections:

```
1. Decrypt the input share for each report share as described in {{input-share-decryption}}.
1. Check that the resulting input share is valid as described in {{input-share-batch-validation}}.
1. Initialize VDAF preparation and initial outputs as described in {{input-share-prep}}.
```

To begin this process, the helper first checks that the nonces in `AggregateInitReq.report_shares`
are all distinct. If two ReportShare values have the same nonce, then the helper MUST abort with
error "unrecognizedMessage". If this check succeeds, the helper then attempts to recover each
input share in `AggregateInitReq.report_shares` as follows:

1. Decrypt the input share for each report share as described in {{input-share-decryption}}.
1. Check that the resulting input share is valid as described in {{input-share-batch-validation}}.

> 1. Initialize VDAF preparation and initial outputs as described in {{input-share-prep}}.

Reviewer 2: For instance, this is common to leader and helper.

Once the helper has processed each valid report share in `AggregateInitReq.report_shares`, the
helper then creates an AggregateInitResp message to complete its initialization. This message is
structured as follows:

~~~
enum {
  continued(0),
  finished(1)
  failed(2),

> } PrepareStepResult;

Author: Only to match VDAF. See https://github.com/cjpatton/vdaf/issues/48.

and `server_role` is 0x02 if the aggregator is the leader and 0x03 otherwise.
If decryption fails, the aggregator marks the report share as invalid with the
error `hpke-decrypt-error`. Otherwise, it outputs the resulting `input_share`.

#### Input Share Validation {#input-share-batch-validation}

Validating an input share will either succeed or fail. In the case of failure,
the input share is marked as invalid with a ReportShareError of type `vdaf-prep-error`.

The validation checks are as follows.

1. Check if the report has never been aggregated but is contained by
   a batch that has been collected. If this check fails, the input share
   is marked as invalid with the error `batch-collected`. This prevents
   additional reports from being aggregated after its batch has already
   been collected.
2. Check if the report has already been aggregated. If this check fails,
   the input share is marked as invalid with the error `report-replayed`.
   This is the case if the report was used in a previous aggregate request
   and is therefore a replay. An aggregator may also choose to mark an
   input share as invalid with the  error `report-dropped` under the conditions

>    prescribed in {{anti-replay}}.

Reviewer 2: Shouldn't the nonce check go here?

Author: Sorry, what nonce check?

Reviewer 2: The check that the nonces aren't internally duplicated.

### Aggregate Continuation {#agg-continue-flow}

In the continuation phase, the leader drives the VDAF preparation of each share in the
candidate report set until the underlying VDAF moves into a terminal state, yielding
an output share for all aggregators or an error. This phase may involve multiple rounds
of interaction depending on the underlying VDAF parameters. Each round trip is initiated
by the leader.

#### Leader Continuation

The leader begins each round of continuation for a report share based on its locally computed
prepare message and the previous PrepareStep from the helper. If PrepareStep is of type "failed",
then the leader marks the report as failed and removes it from the candidate report set and does not
process it further. If the type is "finished", then the leader aborts with "unrecognizedMessage".
[[OPEN ISSUE: This behavior is not specified.]] If the type is "continued", then the leader proceeds as
follows.

Let `leader_outbound` denote the leader's prepare message and `helper_outbound` denote the
helper's. The leader computes the next state transition as follows:

~~~
inbound = VDAF.prep_shares_to_prep(agg_param, [leader_outbound, helper_outbound])

> out = VDAF.prep_next(prep_state, inbound)

Author: I think we had Process at some point in this PR, but that was confusing since it didn't align with the VDAF terminology. I suggest we punt this to the VDAF draft and then just use whatever term makes sense there.

> out = VDAF.prep_next(prep_state, inbound)

Author: https://github.com/cjpatton/vdaf/issues/48

  opaque agg_param<0..2^16-1>;
  opaque helper_state<0..2^16>;
  ReportShare report_shares<1..2^16-1>;
} AggregateInitReq;
~~~

The `nonce` and `extensions` fields of each ReportShare match that in the Report
uploaded by the client. The `encrypted_input_share` field is the `HpkeCiphertext`
whose index in `Report.encrypted_input_shares` is equal to the index of the aggregator
in the task's `aggregator_endpoints`. The `agg_param` field is an opaque, VDAF-specific
aggregation parameter. The `helper_state` parameter contains the helper's state.
This is an optional parameter of an aggregate request that the helper can use to carry
state across requests and across aggregate flows.

Let `[aggregator]` denote the helper's API endpoint. The leader sends a POST
request to `[aggregator]/aggregate` with its AggregateInitReq message as
the payload. The media type is "message/ppm-aggregate-init-req".

#### Helper Initialization

Each helper begins their portion of the aggregate initialization phase with the set
of candidate report shares obtained in an `AggregateInitReq` message from the leader.
It attempts to recover and validate the corresponding input shares similar to the leader,

> and eventually returns a response to the leader carrying the preparation state for each

Reviewer 3: ```suggestion
and eventually returns a response to the leader carrying a VDAF-specific message for each
```

AggregateInitReq. If any message appears out of order, is missing, has an
unrecognized nonce, or if two messages have the same nonce, then the leader MUST
abort with error "unrecognizedMessage".

[[OPEN ISSUE: the leader behavior here is sort of bizarre -- to whom does it abort?]]

#### Input Share Decryption {#input-share-decryption}

Each report share has a corresponding task ID, nonce, list of extensions, and encrypted
input share. Let `nonce`, `extensions`, and `encrypted_input_share` denote these
values, respectively. Given these values, an aggregator decrypts the input
share as follows. First, the aggregator looks up the HPKE config and corresponding
secret key indicated by `encrypted_input_share.config_id`. If not found, then it
marks the report share as invalid. Otherwise, it decrypts the payload with the
following procedure:

~~~
context = SetupBaseR(encrypted_input_share.enc, sk, task_id ||
                     "ppm-00 input share" || 0x01 || server_role)

input_share = context.Open(nonce || extensions,
                           encrypted_input_share.payload)
~~~

where `sk` is the HPKE secret key, `task_id` is the task ID, `nonce` and

> `extensions` are the nonce and extensions of the report share respectively,

Author: Yep, that's right. I left it here to be specific. We can omit in the future if desired.

and `server_role` is 0x02 if the aggregator is the leader and 0x03 otherwise.
If decryption fails, the aggregator marks the report share as invalid. Otherwise,
it outputs the resulting `input_share`.

#### Input Share Batch Validation {#input-share-batch-validation}

Validating an input share will either succeed or fail. In the case of failure,
the input share is marked as invalid with a ReportShareError.

The validation checks are as follows.

1. Check if the report has never been aggregated but is contained by
   a batch that has been collected. If this check fails, the input share
   is marked as invalid with the error `batch-collected`. This prevents
   additional reports from being aggregated after its batch has already
   been collected.
2. Check if the report has already been aggregated. If this check fails,

>    the input share is marked as invalid with the error `report-replayed`.

Author: Let's address this in subsequent changes. We _still_ don't have batch validation requirements nailed down, so this is going to likely change anyway.

   This is the case if the report was used in a previous aggregate request
   and is therefore a replay. An aggregator may also choose to mark an
   input share as invalid with the  error `report-dropped` under the conditions
   prescribed in {{anti-replay}}. The helper also checks to see if the report

>    has never been aggregated but is contained by a batch that has been collected.

Author: Indeed, and it's also helper-specific, so I'll remove it.

If both checks succeed, the input share is not marked as invalid.

#### Input Share Preparation {#input-share-prep}

Input share preparation consists of running the preparation-state initialization
algorithm for the VDAF associated with the task and computes the first state transition.
This produces three possible values:

1. An error, in which case the input report share is marked as invalid.
1. An output share, in which case the aggregator stores the output share for future collection
   as described in {{collect-flow}}.
1. An initial VDAF state and preparation message, denoted `(prep_state, prep_msg)`.

Each aggregator runs this procedure for a given input share with corresponding nonce
as follows:

~~~
prep_state = VDAF.prep_init(vdaf_verify_param, agg_param, nonce, input_share)
out = VDAF.prep_next(prep_state, None)
~~~

`vdaf_verify_param` is the public VDAF parameter, and `agg_param` is the opaque aggregation
parameter. If either step fails, the aggregator marks the report as invalid with error
`vdaf-prep-error`.

> Otherwise, the value `out` is interpreted as follows. If the VDAF is 0-round, then `out`
> is the aggregator's output share. Otherwise, if the VDAF consists of one round or more,
> then the aggregator interprets `out` as the pair `(prep_state, prep_msg)`.

Reviewer: ```suggestion
Otherwise, the value `out` is interpreted as follows. If this is the last round of the VDAF,
then `out` is the aggregator's output share. Otherwise, `out` is the pair `(prep_state, prep_msg)`.
```
To match the corresponding paragraph describing the leader's behavior on L952.

message in `AggregateContinueReq` and its current preparation state (`prep_state`). This
step yields one of three outputs:

1. An error, in which case the input report share is marked as invalid.
1. An output share, in which case the helper stores the output share for future collection
   as described in {{collect-flow}}.
1. An updated VDAF state and preparation message, denoted `(prep_state, prep_msg)`.

To carry out this step, for each PrepareStep in AggregateContinueReq.prepare_shares received
from the leader, the helper performs the following check to determine if the report share
should continue being prepared.

* If failed, then mark the report as failed and reply with a failed PrepareStep
  to the leader.
* If finished, then mark the report as finished and reply with a finished PrepareStep
  to the leader. The helper then stores the output share and awaits for collection;
  see {{collect-flow}}.

Otherwise, preparation continues. In this case, the helper computes its updated state
and output message as follows:

~~~
out = VDAF.prep_next(prep_state, inbound)
~~~

where `inbound` is the previous VDAF preapre message sent by the leader and `prep_state` is

> its current preparation state. If this operation fails, then the helper fails

Reviewer: ```suggestion
the helper's current preparation state. If this operation fails, then the helper fails
```
"its" is ambiguous since the sentence earlier refers to the leader.

with error `vdaf-prep-error`. Otherwise, it interprets `out` as follows. If this
is the last round of VDAF preparation phase, then `out` is the helper's output
share, in which case the helper stores the output share for future collection.
Otherwise, the helper interpets `out` as the tuple `(new_state, prep_msg)`, where
`new_state` is its updated preparation state and `prep_msg` is its next VDAF
message.

This output message for each report in AggregateContinueReq.prepare_shares is then sent
to the leader in an AggregateContinueResp message, structured as follows:

~~~
struct {
  opaque helper_state<0..2^16>;

>   PrepareStep prepare_shares<1..2^16-1>;

Reviewer 3: nit: for consistency with similar abbreviation elsewhere
```suggestion
  PrepareStep prep_shares<1..2^16-1>;
```

Author: Keeping this as-is for now. We can bash the name in a followup change.

} AggregateContinueResp;
~~~

> The order of AggregateContinueResp.prepare_shares matches that of the PrepareStep values in
> `AggregateContinueReq.prepare_shares`. The helper's response to the leader is an HTTP 200 OK whose body

Reviewer: ```suggestion
The order of AggregateContinueResp.prepare_shares MUST match that of the PrepareStep values in
`AggregateContinueReq.prepare_shares`. The helper's response to the leader is an HTTP 200 OK whose body
```

  hpke-unknown-config-id(3),
  hpke-decrypt-error(4),
  vdaf-prep-error(5),
} ReportShareError;
~~~

The leader and helper initialization behavior is detailed below.

#### Leader Initialization

The leader begins the aggregate initialization phase with the set of candidate report
shares as follows:

1. Decrypt the input share for each report share as described in {{input-share-decryption}}.
1. Check that the resulting input share is valid as described in {{input-share-batch-validation}}.
1. Initialize VDAF preparation as described in {{input-share-prep}}.

If any step yields an invalid report share, the leader removes the report share from
the set of candidate reports. Once the leader has initialized this state for all valid
candidate report shares, it then creates an AggregateInitReq message for each helper to
initialize the preparation of this candidate set. The AggregateInitReq message is
structured as follows:

~~~
struct {
  Nonce nonce;
  Extension extensions<0..2^16-1>;
  HpkeCiphertext encrypted_input_share;
} ReportShare;

struct {
  TaskID task_id;
  opaque agg_param<0..2^16-1>;
  opaque helper_state<0..2^16>;
  ReportShare report_shares<1..2^16-1>;
} AggregateInitReq;

> ~~~

Author: ```suggestion
~~~

[[OPEN ISSUE: consider sending report shares separately (in parallel) to the aggregate instructions. RIght now, aggregation parameters and the corresponding report shares are sent at the same time, but this may not be strictly necessary. ]]
```

values, respectively. Given these values, an aggregator decrypts the input
share as follows. First, the aggregator looks up the HPKE config and corresponding
secret key indicated by `encrypted_input_share.config_id`. If not found, then it
marks the report share as invalid with the error `hpke-unknown-config-id`.
Otherwise, it decrypts the payload with the following procedure:

~~~
context = SetupBaseR(encrypted_input_share.enc, sk, task_id ||
                     "ppm-00 input share" || 0x01 || server_role)

input_share = context.Open(nonce || extensions,
                           encrypted_input_share.payload)
~~~

where `sk` is the HPKE secret key, `task_id` is the task ID, `nonce` and
`extensions` are the nonce and extensions of the report share respectively,
and `server_role` is 0x02 if the aggregator is the leader and 0x03 otherwise.
If decryption fails, the aggregator marks the report share as invalid with the
error `hpke-decrypt-error`. Otherwise, it outputs the resulting `input_share`.

#### Input Share Validation {#input-share-batch-validation}

Validating an input share will either succeed or fail. In the case of failure,

> the input share is marked as invalid with a ReportShareError of type `vdaf-prep-error`.

Reviewer 4: I think "vdaf-prep-error" here is incorrect. The two checks below each specify different ReportShareError codes. The next section, "Input Share Preparation", correctly says it may produce a "vdaf-prep-error" error.

Author: Yep, resolved. Thanks!

> the input share is marked as invalid with a ReportShareError of type `vdaf-prep-error`.

Author: ```suggestion
the input share is marked as invalid with a corresponding ReportShareError error.
```

- Initialization: Begin the aggregation flow by sharing report shares with each
  helper. Each aggregator, including the leader, initializates the underlying
  VDAF instance using these report shares and the VDAF configured for the
  corresponding measurement task.
- Continuation: Continue the aggregation flow by exchanging messages produced
  by the underlying VDAF instance until aggregation completes or an error occurs.
  These messages do not replay the shares.
- Completion: Finish the aggregate flow, yielding an output share corresponding
  for each input report share in the batch.

### Aggregate Initialization {#agg-init}

The leader begins aggregation by choosing a set of candidate reports that is subject
to the following restrictions:

* Each report in the set MUST pertain to the same PPM task.
* It is an error to allow a report to be replayed. If the leader has already aggregated
  a report, but the report does not pertain to a batch that has been collected, then
  the report MUST be excluded.
* It is an error to include a new report in a batch that has already been collected.
  If the report pertains to a batch that has been collected, but the leader has not yet

>   aggregated the report, then it MUST be excluded.

Author: Agreed that it's redundant. I think we can omit this text and the forward reference entirely.

of candidate report shares obtained in a `AggregateInitReq` message from the leader.
It attempts to recover and validate the corresponding input shares similar to the leader,
and eventually returns a response to the leader carrying the preparation state for each
report share.

To begin this process, the helper first checks that the nonces in `AggregateInitReq.report_shares`
are all distinct. If two ReportShare values have the same nonce, then the helper MUST abort with
error "unrecognizedMessage". If this check succeeds, the helper then attempts to recover each
input share in `AggregateInitReq.report_shares` as follows:

1. Decrypt the input share for each report share as described in {{input-share-decryption}}.
1. Check that the resulting input share is valid as described in {{input-share-batch-validation}}.
1. Initialize VDAF preparation and initial outputs as described in {{input-share-prep}}.

Once the helper has processed each valid report share in `AggregateInitReq.report_shares`, the
helper then creates an AggregateInitResp message to complete its initialization. This message is
structured as follows:

~~~
enum {
  continued(0),
  finished(1)
  failed(2),

> } PrepareResult;

Author: How about `PrepareStepResult`? (We can bike shed the name later)

struct {
  Nonce nonce;
  PrepareResult prepare_result;
  select (PrepareShare.prepare_result) {
    case continued: opaque prep_msg<0..2^16-1>; // VDAF preparation message
    case finished:  // empty
    case failed:    ReportShareError;
  }

> } PrepareShare;

Author: That works for me, and am happy to bike shed later.

encrypted in order to protect user privacy. However, the details of precisely
how the state is encrypted and the information that it carries is up to the
helper implementation.

[[OPEN ISSUE: we may end up removing helper_state. See #185]]

The rest of the message is a sequence of PrepareShare values, the order of which
matches that of the ReportShare values in `AggregateInitReq.report_shares`. Each report
that was marked as invalid is assigned the PrepareResult `failed`. Otherwise, the
PrepareShare is either marked as continued with the output `prep_msg`, or is marked
as finished if the VDAF computation is finished.

The helper's response to the leader is an HTTP 200 OK whose body is the
AggregateInitResp and media type is "message/ppm-aggregate-init-resp".

Upon receipt of a helper's AggregateInitResp message, the leader checks that the
sequence of PrepareShare messages corresponds to the ReportShare sequence of the
AggregateInitReq. If any message appears out of order, is missing, has an
unrecognized nonce, or if two messages have the same nonce, then the leader MUST
abort with error "unrecognizedMessage".

> [[OPEN ISSUE: the leader behavior here is sort of bizarre -- to whom does it abort?]]

Author: Indeed -- it's above.

it outputs the resulting `input_share`.

#### Input Share Batch Validation {#input-share-batch-validation}

Validating an input share will either succeed or fail. In the case of failure,
the input share is marked as invalid with a ReportShareError.

The validation checks are as follows.

1. Check if the report has never been aggregated but is contained by
   a batch that has been collected. If this check fails, the input share
   is marked as invalid with the error `batch-collected`. This prevents
   additional reports from being aggregated after its batch has already
   been collected.
2. Check if the report has already been aggregated. If this check fails,
   the input share is marked as invalid with the error `report-replayed`.
   This is the case if the report was used in a previous aggregate request
   and is therefore a replay. An aggregator may also choose to mark an
   input share as invalid with the  error `report-dropped` under the conditions
   prescribed in {{anti-replay}}. The helper also checks to see if the report
   has never been aggregated but is contained by a batch that has been collected.

If both checks succeed, the input share is not marked as invalid.

> 

Reviewer 3: Bump.

> 

Author: Let's do this in a separate PR. We still don't have these requirements nailed down and the text is subject to change.

Each aggregator runs this procedure for a given input share with corresponding nonce
as follows:

~~~
prep_state = VDAF.prep_init(vdaf_verify_param, agg_param, nonce, input_share)
out = VDAF.prep_next(prep_state, None)
~~~

`vdaf_verify_param` is the public VDAF parameter, and `agg_param` is the opaque aggregation
parameter. If either step fails, the aggregator marks the report as invalid with error
`vdaf-prep-error`.

Otherwise, the value `out` is interpreted as follows. If the VDAF is 0-round, then `out`
is the aggregator's output share. Otherwise, if the VDAF consists of one round or more,
then the aggregator interprets `out` as the pair `(prep_state, prep_msg)`.

### Aggregate Continuation {#agg-continue-flow}

In the continuation phase, the leader drives the VDAF preparation of each share in the
candidate report set until the underlying VDAF moves into a terminal state, yielding
an output share for all aggregators or an error. This phase may involve multiple rounds
of interaction depending on the underlying VDAF parameters. Each round trip is initiated
by the leader.

#### Leader Continuation

> The leader begins each round of continuation for a report share based on a previous
> response from the helper, denoted `helper_outbound`, and a previous message produced
> from the leader, denoted `leader_outbound`. Each helper message carries a PrepareResult,
> indicating if preparation for the report share should continue, has failed, or is finished.
> 
> If the helper message failed, then the leadaer marks the report as failed and removes it
> from the candidate report set and does not process it further. Otherwise, the leader
> computes its next state transition as follows:

Reviewer 3: ```suggestion
The leader begins each round of continuation for a report share based on its locally computed 
prepare message and the previous PrepareShare from the helper. If PrepareShare is of type "failed", 
then the leader marks the report as failed and removes it from the candidate report set and does not
process it further. If the type is "finished", then the leader aborts with "unrecognizedMessage".
[[OPEN ISSUE: This behavior is not specified.]] If the type is "continued", then the leader proceeds as
follows.

Let `leader_outbound` denote the leader's prepare message and `helper_outbound` denote the
helper's. The leader computes the next state transition as follows:
```

The helper continues with preparation for a report share by combining the leader's input
message in `AggregateContinueReq` and its current preparation state (`prep_state`). This
step yields one of three outputs:

1. An error, in which case the input report share is marked as invalid.
1. An output share, in which case the helper stores the output share for future collection
   as described in {{collect-flow}}.
1. An updated VDAF state and preparation message, denoted `(prep_state, prep_msg)`.

To carry out this step, for each PrepareShare in AggregateContinueReq.prepare_shares received
from the leader, the helper performs the following check to determine if the report share
should continue being prepared.

* If failed, then mark the report as failed and reply with a failed PrepareShare
  to the leader.
* If finished, then mark the report as finished and reply with a finished PrepareShare
  to the leader. The helper then stores the output share and awaits for collection;
  see {{collect-flow}}.

Otherwise, preparation continues. In this case, the helper computes its updated state
and output message as follows:

~~~
out = VDAF.prep_next(prep_state, inbound)
~~~

> where `inbound` is the previous VDAF message sent by the leader and `prep_state` is

Reviewer 3: ```suggestion
where `inbound` is the previous VDAF preapre message sent by the leader and `prep_state` is
```

its current preparation state. If this operation fails, then the helper fails
with error `vdaf-prep-error`. Otherwise, it interprets `out` as follows. If this
is the last round of VDAF preparation phase, then `out` is the helper's output
share, in which case the helper stores the output share for future collection.
Otherwise, the helper interpets `out` as the tuple `(new_state, prep_msg)`, where
`new_state` is its updated preparation state and `prep_msg` is its next VDAF
message.

This output message for each report in AggregateContinueReq.prepare_shares is then sent
to the leader in an AggregateContinueResp message, structured as follows:

~~~
struct {
  opaque helper_state<0..2^16>;
  PrepareShare prepare_shares<1..2^16-1>;
} AggregateContinueResp;
~~~

The order of AggregateContinueResp.prepare_shares matches that of the PrepareShare values in
`AggregateContinueReq.prepare_shares`. The helper's response to the leader is an HTTP 200 OK whose body
is the AggregateContinueResp and media type is "message/ppm-aggregate-continue-resp". The helper
then awaits the next message from the leader.

> [[OPEN ISSUE: consider relaxing this ordering constraint]]

Reviewer 3: ```suggestion
[[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]
```

be cryptographically protected as described in {{agg-init}}.

~~~
Leader                                                 Helper

Aggregate request (Reports 1-10) ------------------------>  \
<----------------------------- Aggregate response (State 1) | Reports
Aggregate request (continued, State 1) ------------------>  | 1-10
<----------------------------- Aggregate response (State 1) /

Aggregate request (Reports 11-20, State 1) -------------->  \
<----------------------------- Aggregate response (State 1) | Reports
Aggregate request (continued, State 2) ------------------>  | 11-20
<----------------------------- Aggregate response (State 2) /
~~~
{: #aggregation-flow-illustration title="Aggregation Flow (batch size=20).
Multiple aggregation flows can be executed at the same time."}

[OPEN ISSUE: Should there be an indication of whether a given aggregate request is a
continuation of a previous sub-batch?]

The aggregation flow can be thought of as having three phases:

- Initialization: Begin the aggregation flow by sharing report shares with each
  helper. Each aggregator, including the leader, initializates the underlying

>   VDAF instance using these report shares and the public VDAF configured for

Reviewer: What is a "public VDAF"?

Author: Strange leftover. Deleting "public".

  the corresponding measurement task.
- Continuation: Continue the aggregation flow by exchanging messages produced
  by the underlying VDAF instance until aggregation completes or an error occurs.
- Completion: Finish the aggregate flow, yielding an aggregate share corresponding

>   to all output shares in the batch.

Author: Yep, good catch. This should be "output shares."

Once both aggregation is complete and a sufficient number of reports have
been aggregated, the aggregate shares can be collected to produce the final,

> aggregate output. That process is described in {{collect-flow}}.

Author: I think we can remove it, since this is prematurely describing the collect flow.

### Aggregate Initialization {#agg-init}

The leader begins aggregation by choosing a set of candidate reports that is subject to the following restrictions:
* Each report in the set MUST pertain to the same PPM task.
* It is an error to allow a report to be replayed. If the leader has already aggregated a report, but the report does not pertain to a batch that has been collected, then the report MUST be excluded.
* It is an error to include a new report in a batch that has already been collected. If the report pertains to a batch that has been collected, but the leader has not yet aggregated the report, then it MUST be excluded.

After choosing the set of candidates, the The leader begins aggregation by splitting each report into "report
shares", one for each aggregator. A single report share is encoded as follows:

~~~
struct {
  Nonce nonce;
  Extension extensions<0..2^16-1>;
  HpkeCiphertext encrypted_input_share;

> } ReportShare;

Author: Yeah, I think you're right. I'll move it later.

~~~

Once the set of report shares is identified, the leader decrypts its own input
shares and computes the initial preparation state. To do this, the leader starts
by looking up the HPKE config and corresponding secret key indicated by `encrypted_input_share.config_id`.
If not found, then it MUST remove the report share from the set. Otherwise, it
decrypts the payload with the following procedure:

~~~
context = SetupBaseR(encrypted_input_share.enc, sk, task_id ||
                     "ppm-00 input share" || 0x01 || 0x02)

input_share = context.Open(ReportShare.nonce || ReportShare.extensions,
                           ReportShare.encrypted_input_share.payload)
~~~

where `sk` is the HPKE secret key, `task_id` is the task ID, and `nonce` and
`extensions` are the nonce and extensions of the report share respectively.
If decryption fails, the leader marks this report as invalid and removes it
from the set of candidate reports. Moreover, if the report has never been aggregated
but is contained by a batch that has been collected, then the aggregator marks
this report as invalid and removes it from the set of candidate reports.

> Next, the leader runs the preparation-state initialization algorithm for the

Author: That's right 👍 it would be pointless for the leader to kick off aggregation for any share that it would eventually throw away.

VDAF associated with the task and computes the first state transition. Let
`vdaf_verify_param` denote the pre-configured VDAF verification parameter,
`agg_param` denote the aggregation parameter, `nonce` denote the report nonce,
and let `input_share` denote the input share:

~~~
prep_state = VDAF.prep_init(vdaf_verify_param, agg_param, nonce, input_share)
~~~

Once the leader has initialized this state for all candidate report shares, it
then creates an AggregateInitReq message for each helper to initialize the
process of aggregating this candidate set. This message is structured as follows:

~~~
struct {
  TaskID task_id;
  opaque agg_param<0..2^16-1>;
  opaque helper_state<0..2^16>;

>   ReportShare seq<1..2^16-1>;

Reviewer 2: Can we rename this to "reports"?

Author: That's better!

} AggregateInitReq;
~~~

The `nonce` and `extensions` fields of each ReportShare match that in the Report
uploaded by the client. The `encrypted_input_share` field is the `HpkeCiphertext`
whose index in `Report.encrypted_input_shares` is equal to the index of the aggregator
in the task's `aggregator_endpoints`. The `agg_param` field is an opaque, VDAF-specific
aggregation parameter. The `helper_state` parameter contains the helper's state.
This is an optional parameter of an aggregate request that the helper can use to carry
state across requests and across aggregate flows.

Let `[aggregator]` denote the helper's API endpoint. The leader sends a POST
request to `[aggregator]/aggregate` with its AggregateInitReq message as
the payload. The media type is "message/ppm-aggregate-init-req".

To process the leader's request, the helper first checks that the nonces in
`AggregateInitReq.seq` are all distinct. If two InputReportShare messages have the
same nonce, then the helper MUST abort with error "unrecognizedMessage".

> Next, the helper initializates the preparation state for each report share in

Reviewer 2: How does it do that? Show the API call.

Author: This is shown below.

`AggregateInitReq.seq`. Each report share will either be processed successfully
or is otherwise invalid, being marked with one of the following errors:

~~~
enum {
  batch-collected(0),
  report-replayed(1),
  report-dropped(2),
  hpke-unknown-config-id(3),
  hpke-decrypt-error(4),
  vdaf-prep-error(5),
} ReportShareError;
~~~

Processing a report share begins by first checking if the report has already
been aggregated. This is the case if the report was used in a previous aggregate
request and is therefore a replay. In this case, the helper marks the report as

> invalid with the `report-replayed` error. Note that detecting whether a report

Reviewer 2: Can we just merge this with the uniquness check above?

uncollected batch intervals. So that the aggregator does not have to maintain
this storage indefinitely, it MAY instead mark the report invalid with `report-dropped`
under the conditions prescribed in {{anti-replay}}. The helper also checks to see
if the report has never been aggregated but is contained by a batch that has been
collected. If this is true, then the helper marks the report as invalid with error
`batch-collected`. This prevents additional reports from being aggregated after its
batch has already been collected.

Next, the helper attempts to decrypt its input share. It starts by looking
up the HPKE config and corresponding secret key indicated by
`encrypted_input_share.config_id`. If not found, then it marks the report
as invalid with error `hpke-unknown-config-id`. Otherwise, it decrypts the
payload with the following procedure:

~~~
context = SetupBaseR(encrypted_input_share.enc, sk, task_id ||
                     "ppm-00 input share" || 0x01 || 0x03)

input_share = context.Open(ReportShare.nonce || ReportShare.extensions,
                           ReportShare.encrypted_input_share.payload)

> ~~~

Reviewer 2: See above about merging this.

where `sk` is the HPKE secret key, `task_id` is the task ID, and `nonce` and
`extensions` are the nonce and extensions of the report share respectively.
If decryption fails, then the aggregator marks the report as invalid with
error `hpke-decrypt-error`.

Next, the helper runs the preparation-state initialization algorithm for the
VDAF associated with the task and computes the first state transition:

~~~
prep_state = VDAF.prep_init(vdaf_verify_param, agg_param, nonce, input_share)
~~~

If this step fails, the helper marks the report as invalid with error
`vdaf-prep-error`. Otherwise, the helper produces an initial outbound message:

~~~
out = VDAF.prep_next(prep_state, None)
~~~

The value `out` is interpreted as follows. If the VDAF is 0-round, then `out` is the
aggregator's output share, in which case the aggregator finishes and stores its
output share for further processing as described in {{agg-complete}}. Otherwise,
if the VDAF consists of one round or more, then the aggregator interprets `out` as

> the pair `(new_state, agg_msg)`, where `new_state` is its updated state and `agg_msg`

Author: Yep, that's better!

is its next VDAF message. For the latter case, the helper sets `prep_state` to `new_state`.

Once the helper has processed each report share in `AggregateInitReq.seq`, the helper
then creates an AggregateInitResp message to complete its initialization. This message is
structured as follows:

~~~
enum {
  continued(0),
  finished(1)
  failed(2),
} ProcessType;

struct {
  Nonce nonce;
  ProcessType process_type;
  select (ProcessShare.process_type) {
    case continued: opaque agg_msg<0..2^16-1>; // VDAF output message
    case finished:  // empty
    case failed:    ReportShareError;
  }
} ProcessShare;

struct {
  opaque helper_state<0..2^16>;
  ProcessShare seq<1..2^16-1>;
} AggregateInitResp;
~~~

> The `helper_state` parameter contains the helper's initial state. This is an

Reviewer 2: This section has gotten very long.

Author: Headers included!

state across requests. At least part of the state will usually need to be
encrypted in order to protect user privacy. However, the details of precisely
how the state is encrypted and the information that it carries is up to the
helper implementation.

[[OPEN ISSUE: we may end up removing helper_state. See #185]]

The rest of the message is a sequence of ProcessShare values, the order of which
matches that of the InputReportShare values in `AggregateInitReq.seq`. Each report
that was marked as invalid is assigned the ProcessType `failed`. Otherwise, the
ProcessShare is either marked as continued with the output `agg_msg`, or is marked
as finished if the VDAF computation is finished.

The helper's response to the leader is an HTTP 200 OK whose body is the
AggregateInitResp and media type is "message/ppm-aggregate-init-resp".

Upon receipt of a helper's AggregateInitResp message, the leader checks that the
sequence of ProcessShare messages corresponds to the ReportShare sequence of the
AggregateInitReq. If any message appears out of order, is missing, has an
unrecognized nonce, or if two messages have the same nonce, then the leader MUST

> abort with error "unrecognizedMessage".

Reviewer 2: Why are we both having order and the nonce? Seems like a great opportunity for things to go wrong.

Author: I think we can probably relax this. Let's address it in a followup issue.

Reviewer 5: Followup issue: https://github.com/ietf-wg-ppm/ppm-specification/issues/217 (FWIW, I agree this can be relaxed, I included some potential considerations in the issue.)

### Aggregate Continuation {#agg-continue-flow}

In the continuation phase, the leader drives aggregation of each share in the candidate
report set until the underlying VDAF moves into a terminal state, which happens either
when aggregation completes or an error occurs. The leader does this by repeating the
following procedure.

* If the helper failed, then mark the report as failed, remove it from the candidate
  report set, and do not process it further.

> * If the leader continued but the helper finished, or if the leader finished but

Reviewer 2: Can we get a forward ref to how things ifnish.

Author: Yep, done.

  the helper continued, then mark the report as failed and do not process it further.
* If the leader and helper finished, then continuation is complete. At this point the

>   report has been completely processed and both aggregators have recovered an

Author: Yeah, this is a good suggestion. Let's converge on "prepare" as the verb of choice here.

  output share. The leader proceeds as described in {{agg-complete}}.
* If the leader and helper continued, then the leader computes its next state

>   transition ("prep_next") and sends the helper a ProcessShare message with the

Reviewer 2: Where is this defined?

  payload computed as described below.

The leader sends each helper a ProcessShare value for each report share that is to
be continued. Let `leader_outbound` denote the last VDAF message the leader computed
and let `helper_outbound` denote the last VDAF message it received from the helper.
The payload of the the ProcessShare message is computed as

~~~
inbound = VDAF.prep_shares_to_prep(agg_param, [leader_outbound, helper_outbound])
~~~

where [leader_outbound, helper_outbound] is a vector of two elements. The next state
transition is then computed as:

~~~
out = VDAF.prep_next(prep_state, inbound)
~~~

> If either of these operations fails, then the leader marks the report as invalid.

Reviewer 2: If it's invalid does it have to tell the helper?

Reviewer 3: Our thinking here is that the leader would just skip reports that can't be processed further. The motivation for this is to keep the helper simple. It also has the advantage of avoiding communication overhead for reports that can't be aggregated. One potential downside is that the helper doesn't get to find out *why* processing failed --- VDAF evaluation may fail in different ways, for example --- whereas the leader does.

Author: Right now, they shouldn't actually aggregate until the leader issues an AggregateShareReq. We can change how that's done if we want to optimize things, but let's do so separately.

Otherwise it interprets `out` as follows. If this is the last round of VDAF
preparation phase, then `out` is the leader's output share, in which case the leader
finishes aggregation. Otherwise, it interprets `out` as the tuple `(new_state, outbound)`,
where `new_state` is its new preparation state and `outbound` is its next VDAF messaage,
and continues with `outbound` as its next VDAF message. The leader sets `prep_state`
to `new_state`.

The leader then sends each ProcessShare to the helper in an AggregateContinueReq message,
structured as follows:

~~~
struct {
  opaque helper_state<0..2^16>;
  ProcessShare seq<1..2^16-1>;
} AggregateContinueReq;
~~~

For each aggregator endpoint `[aggregator]` in `AggregateContinueReq.task_id`'s
parameters except its own, the leader sends a POST request to
`[aggregator]/aggregate` with AggregateContinueReq as the payload and the media
type set to "message/ppm-aggregate-continue-req".

For each ProcessShare in AggregateContinueReq.seq received from the leader, the helper

> proceeds as follows:

Author: It's "do this check and then proceed accordingly to either continue processing, invalidate the report, or finish."

is the last round of VDAF preparation phase, then `out` is the helper's output
share, in which case the helper finishes and transitions to FINISHED. Otherwise,
the helper interpets `out` as the tuple `(new_state, agg_msg)`, where
`new_state` is its updated preperation state and `agg_msg` is its next VDAF
message.

This output message for each report in AggregateContinueReq.seq is then sent to the leader
in an AggregateContinueResp message, structured as follows:

~~~
struct {
  opaque helper_state<0..2^16>;
  ProcessShare seq<1..2^16-1>;
} AggregateContinueResp;
~~~

The order of AggregateContinueResp.seq matches that of the InputReportShare values in
`AggregateContinueReq.seq`. The helper's response to the leader is an HTTP 200 OK whose body
is the AggregateInitResp and media type is "message/ppm-aggregate-continue-resp". The helper
then awaits the next message from the leader.

### Aggregate Completion {#agg-complete}

Once processing of a report share is finished, each aggregator stores the the
recovered output shares until the batch to which it pertains is collected.
To aggregate the output shares, denoted `out_shares`, the aggregator runs

> the aggregation algorithm specified by the VDAF:

Author: > "output share" should be singular to agree with "report share" on the line above, right?

Yep!

> I've argued before we should spell out the state machine for report shares and for batches. That shouldn't happen in this change, but since a lot of subtle stuff like anti-replay depends on the state of a report vs. the state of a batch, the document would benefit from more clarity about this.

I don't think we have a clear set of requirements for batch validation yet (https://github.com/ietf-wg-ppm/ppm-specification/issues/195), but once we do, yes, we should definitely add more detail.

~~~
agg_share = VDAF.out_shares_to_agg_share(agg_param, out_shares)
~~~

Note that for most VDAFs, it is possible to aggregate output shares as they
arrive rather than wait until the batch is collected. To do so however, it is
necessary to enforce the batch parameters as described in
{{batch-parameter-validation}} so that the aggregator knows which aggregate

> share to update.

Reviewer: This is kind of awkward. The implication is that batch parameters don't need to be validated when using a VDAF without an aggregation parameter. Also I don't understand how validation of batch parameters informs which aggregate is used?

header field to suggest a pulling interval to the collector.

Once all the necessary reports have been prepared, the leader obtains the
helper's encrypted aggregate share for the batch interval by requesting an
aggregate share from each helper.

To do this, the leader first computes a checksum over the set of output shares
included in the batch window. The checksum is computed by taking the SHA256 hash of
each nonce from the client reports included in the aggregation, then combining the

> hash values with a bitwise-XOR operation.

Reviewer: This is identical to the definition of `checksum`, a couple paragraphs later. We should delete one.

Then, for each aggregator endpoint `[aggregator]` in the parameters associated
with `CollectReq.task_id` (see {{collect-flow}}) except its own, the leader sends
a POST request to `[aggregator]/aggregate_share` with the following message:

~~~
struct {
  TaskID task_id;
  Interval batch_interval;
  uint64 report_count;
  opaque checksum[32];
  opaque helper_state<0..2^16>;
} AggregateShareReq;
~~~

* `task_id` is the task ID associated with the PPM parameters.
* `batch_interval` is the batch interval of the request.
* `report_count` is the number of reports included in the aggregation.
* `checksum` is the checksum computed over the set of client reports. The
  checksum is computed by taking the SHA256 hash of each nonce from the client
  reports included in the aggregation, then combining the hash values with a

>   bitwise-XOR operation.

Reviewer 2: Why are we including the nonces and not the values?

* `helper_state` is the helper's state, which is carried across requests from
  the leader.

To handle the leader's request, the helper first ensures that the request meets
the requirements for batch parameters following the procedure in
{{batch-parameter-validation}}. If the batch parameters are invalid, then it
MUST abort with error "batchInvalid".

Next, it computes a checksum based on its view of the output shares included in
the batch window, and checks that the `report_count` and `checksum` included in
the request match its computed values. If not, then it MUST abort with error
"batchMismatch".

Next, it computes the aggregate share for the batch interval as described in
{{agg-complete}}, obtaining an opaque byte string `agg_share` whose
structure is specific to the VDAF in use. The helper encrypts the
`agg_share` as follows:

~~~
enc, context = SetupBaseS(pk, AggregateShareReq.task_id ||
                              "ppm-00 aggregate share" || 0x03 || 0x00)

encrypted_agg_share = context.Seal(AggregateShareReq.batch_interval,

>                                    agg_share)

Reviewer 2: Here too, let's break out the crypto.

~~~

where `pk` is the HPKE public key encoded by the collector's HPKE key.

Encryption prevents the leader from learning the actual result, as it only has
its own aggregate share and cannot compute the helper's.

The helper responds to the leader with HTTP status 200 OK and a body consisting
of an `AggregateShareResp`:

~~~
struct {
    HpkeCiphertext encrypted_aggregate_share;
} AggregateShareResp;
~~~

`encrypted_aggregate_share.config_id` is set to the collector's HPKE config ID.
`encrypted_aggregate_share.enc` is set to the encapsulated HPKE context `enc`
computed above and `encrypted_aggregate_share.ciphertext` is the ciphertext
`encrypted_agg_share` computed above.

After receiving the helper's response, the leader uses the HpkeCiphertext to
respond to a collect request (see {{collect-flow}}).

> The leader MAY make multiple aggregate-share requests for a given batch interval
> and aggregation parameter MUST get the same result each time.

Reviewer: ```suggestion
The leader MAY make multiple aggregate-share requests for a given batch interval
and aggregation parameter and MUST get the same result each time.
```
The intent here is that the helper should use the same set of reports to service multiple aggregate-share requests, right? Could that be stated explicitly.

Author: The intent here is that if the parameters of the aggregation request are the same then the output should be the same.

with `CollectReq.task_id` (see {{collect-flow}}) except its own, the leader sends
a POST request to `[aggregator]/aggregate_share` with the following message:

~~~
struct {
  TaskID task_id;
  Interval batch_interval;
  uint64 report_count;
  opaque checksum[32];
  opaque helper_state<0..2^16>;
} AggregateShareReq;
~~~

* `task_id` is the task ID associated with the PPM parameters.
* `batch_interval` is the batch interval of the request.
* `report_count` is the number of reports included in the aggregation.
* `checksum` is the checksum computed over the set of client reports.
* `helper_state` is the helper's state, which is carried across requests from
  the leader.

To handle the leader's request, the helper first ensures that the request meets
the requirements for batch parameters following the procedure in
{{batch-parameter-validation}}. If the batch parameters are invalid, then it
MUST abort with error "batchInvalid".

Next, it computes a checksum based on its view of the output shares included in
the batch window, and checks that the `report_count` and `checksum` included in
the request match its computed values. If not, then it MUST abort with error

> "batchMismatch".

Reviewer 4: Should we add this to the errors table above? (I see it says "this list is not exhaustive", but I'm not sure if that means other errors may be in the same URI namespace, or other errors would have entirely different URIs)

Reviewer 2: :+1:

Author: Yep, done.

Next, it computes the aggregate share `agg_share` corresponding to the set of
output shares, denoted `out_shares`, for the batch interval, as follows:

~~~
agg_share = VDAF.out_shares_to_agg_share(agg_param, out_shares)
~~~

Note that for most VDAFs, it is possible to aggregate output shares as they
arrive rather than wait until the batch is collected. To do so however, it is
necessary to enforce the batch parameters as described in
{{batch-parameter-validation}} so that the aggregator knows which aggregate

> share to update.

Reviewer: We're discussing adding a couple of methods to VDAF here: https://github.com/cjpatton/vdaf/issues/47

> share to update.

Author: Cool! I'll resolve this then.

output shares, denoted `out_shares`, for the batch interval, as follows:

~~~
agg_share = VDAF.out_shares_to_agg_share(agg_param, out_shares)
~~~

Note that for most VDAFs, it is possible to aggregate output shares as they
arrive rather than wait until the batch is collected. To do so however, it is
necessary to enforce the batch parameters as described in
{{batch-parameter-validation}} so that the aggregator knows which aggregate
share to update.

The helper then encrypts `agg_share` under the collector's HPKE public
key as described in {{aggregate-share-encrypt}}, yielding `encrypted_agg_share`.
Encryption prevents the leader from learning the actual result, as it only has
its own aggregate share and cannot compute the helper's.

The helper responds to the leader with HTTP status 200 OK and a body consisting
of an `AggregateShareResp`:

~~~
struct {
    HpkeCiphertext encrypted_aggregate_share;
} AggregateShareResp;
~~~

`encrypted_aggregate_share.config_id` is set to the collector's HPKE config ID.
`encrypted_aggregate_share.enc` is set to the encapsulated HPKE context `enc`
computed above and `encrypted_aggregate_share.ciphertext` is the ciphertext

> `encrypted_agg_share` computed above.

Reviewer: I think this discussion of how to populate the fields of `encrypted_aggregate_share` would be more clear in the `aggregate-share-encrypt` section, below.

After receiving the helper's response, the leader uses the HpkeCiphertext to
respond to a collect request (see {{collect-flow}}).

The leader MAY make multiple aggregate-share requests for a given batch interval

> and aggregation parameter and MUST get the same result each time.

Reviewer 3: I don't understand why this behaviour is useful. Is this new, or is this already in the spec?

Reviewer: I'm not sure if this is what Chris W. had in mind but I agree with this text. Suppose we have a task with `max_batch_lifetime = 1`. Consider:

1. Leader makes an `AggregateShareReq` for some batch interval.
2. Helper services the request, and marks that batch interval as having been collected once.
3. Helper attempts to transmit an `AggregateShareResp` to leader, but the message is truncated.

Now, if the leader retries its `AggregateShareReq`, the helper will refuse the request because the batch interval's lifetime has been consumed. So, to allow the leader to retry this request, the helper has to be willing to resend previously computed aggregate shares.

Author: Yanking this out to discuss in a separate issue: https://github.com/ietf-wg-ppm/ppm-specification/issues/226

share `agg_share` by aggregating all of the prepared output shares that fall within
the batch interval. Finally, it encrypts it under the collector's HPKE public key as
described in {{aggregate-share-encrypt}}.

### Collection Finalization {#collect-finalization}

Once the collector has received a successful collect response from the leader,
it can decrypt the aggregate shares and produce an aggregate result. The collector
decrypts each aggregate share as described in {{aggregate-share-encrypt}}.
If the collector successfully decrypts all aggregate shares, the collector
then unshards the aggregate shares into an aggregate result using the VDAF's
`agg_shares_to_result` algorithm. In particular, let `agg_shares` denote the
ordered sequence of aggregator shares, ordered by aggregator index, and let
`agg_param` be the opaque aggregation parameter. The final aggregate result
is computed as follows:

~~~
agg_result = VDAF.agg_shares_to_result(agg_param, agg_shares)
~~~

### Aggregate Share Encryption {#aggregate-share-encrypt}

Encrypting an aggregate share `agg_share` for a given `AggregateShareReq` is done
as follows:

~~~
enc, context = SetupBaseS(pk, AggregateShareReq.task_id ||

>                               "ppm-00 aggregate share" || 0x03 || 0x00)

Reviewer 3: Replac
```suggestion
                              "ppm-00 aggregate share" || server_role || 0x00)
```
where `server_role` is `0x02` for the leader and `0x03` for a helper.
